608 add threshold weighted scores

nicholasloveday opened the pull request:
Please work through the following checklists. Delete anything that isn't relevant.
## Development for new xarray-based metrics
- [x] Works with n-dimensional data and includes `reduce_dims`, `preserve_dims`, and `weights` args.
- [x] Typehints added
- [x] Docstrings complete and followed Napoleon (google) style
- [x] Reference to paper/webpage is in docstring
- [x] Add error handling
- [x] Imported into the API

## Testing of new xarray-based metrics
- [x] 100% unit test coverage
- [x] Test that metric is compatible with dask.
- [x] Test that metrics work with inputs that contain NaNs
- [x] Test that broadcasting with xarray works
- [x] Test both reduce and preserve dims arguments work
- [x] Test that errors are raised as expected

## Tutorial notebook 
- [x] Short introduction to why you would use that metric and what it tells you
- [x] A link to a reference
- [x] A "things to try next" section at the end
- [x] Add notebook to [Tutorial_Gallery.ipynb](https://github.com/nci/scores/blob/develop/tutorials/Tutorial_Gallery.ipynb)
- [x] Optional - a detailed discussion of how the metric works at the end of the notebook

## Documentation
- [x] Add the score to the [API documentation](https://github.com/nci/scores/blob/develop/docs/api.md)
- [x] Add the score to the [included list of metrics and tools](https://github.com/nci/scores/blob/develop/docs/included.md)

Steph-Chong commented on docs/included.md:
```diff
@@ -117,6 +117,22 @@
   - [API](api.md#scores.continuous.rmse)
   - [Tutorial](project:./tutorials/Root_Mean_Squared_Error.md)
   - [Wikipedia](https://en.wikipedia.org/wiki/Root-mean-square_deviation)
+* - Threshold weighted score
```
```suggestion
* - Threshold Weighted Score
```
For consistency of capitalisation with the other entire in the table

Steph-Chong commented on docs/included.md:
```diff
@@ -117,6 +117,22 @@
+  - [API](api.md#scores.continuous.threshold_weighted_score)
+  - &mdash;
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+  * - Threshold weighted squared error
```
```suggestion
  * - Threshold Weighted Squared Error
```

Steph-Chong commented on docs/included.md:
```diff
@@ -117,6 +117,22 @@
+  - [API](api.md#scores.continuous.threshold_weighted_squared_error)
+  - &mdash;
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+  * - Threshold weighted absolute error
```
```suggestion
  * - Threshold Weighted Absolute Error
```

Steph-Chong commented on docs/included.md:
```diff
@@ -117,6 +117,22 @@
+  - [API](api.md#scores.continuous.threshold_weighted_absolute_error)
+  - &mdash;
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+  * - Threshold weighted quantile score
```
```suggestion
  * - Threshold Weighted Quantile Score
```

Steph-Chong reviewed (commented):
Some minor feedback.

In included.md:
1. Capitalisation (for consistency with other table entries)
2. The four entries don't appear to be in alphabetical order 

In the docstrings, the references don't appear to follow APA7

tennlee commented on src/scores/continuous/__init__.py:
```diff
@@ -30,7 +36,7 @@
     "murphy_thetas",
     "quantile_score",
-    "correlation",
+    "pearsonr",
```
I think this might not be quite right. 'pearsonr' now sits in 'scores.continuous.correlation.pearsonr'. The 'correlation' in __all__ isn't causing any problem that I'm aware of, but presumably points to the correlation module whereas pre the last release was a score. I don't think the __all__ refer should point to pearsonr.

nicholasloveday replied:
Ah yes - I have reverted this.

tennlee commented:
I haven't reviewed the tests yet, but so I don't neglect to mention it later, I will be looking to see whether there are some consistency checks against unweighted scores when the weights are set neutrally.

nicholasloveday commented:
> I haven't reviewed the tests yet, but so I don't neglect to mention it later, I will be looking to see whether there are some consistency checks against unweighted scores when the weights are set neutrally.

You will find these consistency tests in `test_threshold_weighted_score1`. Huber Loss isn't in scores yet, so I copied the Huber Loss output from an independent calculation.

nicholasloveday commented:
> Some minor feedback.
> 
> In included.md:
> 
> 1. Capitalisation suggestions (for consistency with other table entries)
> 2. The four entries don't appear to be in alphabetical order
> 
> In the docstrings, the references don't appear to follow APA7.
> 
> Edit: Re. API docstrings rendering in readthedocs - I think you might need two back ticks (not just one) around function names. (You can check when it is put into branch 228, or if you build it locally in readthedocs). Further Edit: it looks like sometimes you need one back tick and sometimes you need two back ticks. It looks like @nikeethr used a mix (depending on use case) in FSS API docstrings, and the FSS API docstrings appear to be rendering nicely in readthedocs.

Thanks @Steph-Chong . I think I have addressed this now

nicholasloveday commented:
This pull request is now ready for review.

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,657 @@
+def threshold_weighted_score(
+    fcst: xr.DataArray,
+    obs: xr.DataArray,
+    scoring_func: Literal[
```
Is the best as a string, as a function pointer, or as multiple  functions (e.g. threshold_weighted_squared_error, threshold_weighted_absolute_error), or as a base class with classes. e.g. class TW_SquaredError(ThresholdWeightedScore).  I'm not saying the current approach is wrong, but whenever I see strings being passed as selectors I think it's worth asking the question. Or, potentially, scores.continuous.tw.squared_error, scores.continuous.tw.huber_loss etc.

nicholasloveday replied:
A couple of reasons for a function rather than a class:
1. It follows the pattern of all the metrics in scores (the exception is the contingency table class which allows multiple metrics to be calculated from a contingency table).
2. Backwards compatibility with Jive.

I have some convenience functions towards the end of this file that align with your last suggestion. More could be added in the future.

tennlee replied:
For clarity, the downsides (mainly to API users) of using strings in this fashion are as follows:

- Unlike a function name or a function pointer, code completion and static analysis doesn't understand what's in the string
- The API is not as self-documenting / discoverable in terms of having clear API endpoints for each supported threshold-weightable function
- Things like typos in the strings aren't going to be picked up with a clear warning message as would be the case  with a mistyped function name or a mistyped function argument
- Inconsistent with other approaches in the codebase - e.g. the consist_impl files has e.g. consistent_quantile_score, consistent_huber_score rather than a general consistenct_sore(score_type='huber')
- Ability to separately provide docstrings for each threshold-weighted score which describe that score specifically
- Allows test functions to be more specific to just one score type
- It's not possible to pass in a novel score which passes custom values to consistent_expectile_score in a new way. While users could call consistent_expectile_score directly, they would also have to then take on the work of checking arguments and extracting g, phi and phi_prime, resulting in potential inconsistencies.

On option would be to make threshold_weighted_score a private method (i.e. not present in the public API) and only make the convenience functions part of the public API. 

The actual functionality in threshold_weighted_score appears to be to check the arguments, and extract g, phi and phi_prime. The scoring func could be a callable rather than a string, perhaps defined as functions in the module. If threshold_weighted_score is mainly a worker engine for the convenience functions, then the arguments against using a string selector are even stronger. 

If any of those suggestions seems reasonable, I'm happy to refactor the code and put that up for you to consider.

I also have just had a lot of problems with string-passing code generally, which doesn't directly apply here but has left me generally with the view that using strings to control logic flows of what's inside a function is likely to make the code harder to understand.

I can be convinced differently, but I think my perspective would be to make the convenience functions the primary recommended API for users, put them in a scores.continuous.tw module, probably also prefix the methods with e.e.g tw_absolute_error, tw_quantile_score so it's still clear if people are just using the score name, keep the general threshold_weights_score method as an underlying engine, but pass in callables rather than strings.

tennlee replied:
For completeness, I'm totally fine with not using classes here, I'm just talking about making the most elegant API for users which isn't too brittle / error-prone.

nicholasloveday replied:
My plan is:
- Get rid of `threshold_weighted_score`. There will need to be some updates in Jive.
- Have convenience functions (6 of them) instead. `_check_tws_args` and `_auxiliary_funcs` is called in each one.
- Move the second two checks to the relevant functions.

This will remove strings from being passed around in all places.

I'll push up a commit where this has been done for tw_squared_error with tests passing, but will continue on with refactoring the rest of the scores until I hear otherwise.

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,687 @@
+    result = (
+        result1.where(x >= a, result0)
+        .where(x < b, result2)
+        .where(x < c, result3)
```
Just a suggestion, easier to debug by splitting this out into different lines rather than using chaining.

nicholasloveday replied:
Removed chaining

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,687 @@
+    result = (
+        result1.where(x >= a, result0)
+        .where(x < b, result2)
+        .where(x < c, result3)
```
Just a suggestion, easier to debug by splitting this out into different lines rather than using chaining.

nicholasloveday replied:
Removed chaining

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,687 @@
+
+
+def _check_tws_args(
+    scoring_func=None, alpha=None, huber_param=None, interval_where_one=None, interval_where_positive=None
```
I also just noticed the lack of type hints

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,687 @@
+    _check_tws_args(alpha=0.5, interval_where_one=interval_where_one, interval_where_positive=interval_where_positive)
+    g, _, _ = _auxiliary_funcs(fcst, obs, interval_where_one, interval_where_positive)
+
+    return 2 * consistent_quantile_score(
```
A suggestion only. Sometimes it can make sense to replace the numbers (here 2 and 0.5) with a semantic variable. i.e. something like assigning even_weighting=0.5 then using that in the call to consistent_quantile_score. It can help make code more readable. That said, this particular code is really something that takes a specialist to follow and has been tested, so it's up to you if you think it's going to help make it more readable/maintainable.

nicholasloveday replied:
I have added comments. I think it will be easier to read if you can see exactly what the scaling factor is rather than defining it elsewhere.

Note that the scaling function for `tw_expectile_score` now differs from what you previously saw.

The calculations of $\phi$ and $\phi'$ in `_auxiliary_funcs` use $\phi(s)=2s^2$. This means that for `tw_squared_error` requires no scaling, but  `tw_expectile_score` which may seem counterintuitive. I've added comments to describe this while maintaining the code to be consistent with the literature.

tennlee commented on docs/included.md:
```diff
@@ -117,6 +117,26 @@
   - [API](api.md#scores.continuous.rmse)
   - [Tutorial](project:./tutorials/Root_Mean_Squared_Error.md)
   - [Wikipedia](https://en.wikipedia.org/wiki/Root-mean-square_deviation)
+* - Threshold Weighted Absolute Error
```
The docs build process is giving me an error. It says

docs/included.md:5: ERROR: Error parsing content block for the "list-table" directive: two-level bullet list expected, but row 25 does not contain a second-level bullet list

Steph-Chong replied:
I think I may have found the errors in the list table.

It looks like there is an indentation in front of the asterisk for each of:
- Threshold Weighted Expectile Score
- Threshold Weighted Huber Loss
- Threshold Weighted Squared Error
- Threshold Weighted Quantile Score

Threshold Weighted Absolute Error looked okay to me

nicholasloveday replied:
Thanks for this pick up - hopefully I have fixed it

tennlee commented on tutorials/Threshold_Weights_Scores.ipynb:
I get an import error: cannot import name 'threshold_weighted_score" from 'scores.continuous'. Please re-run the notebooks and confirm they are working after the latest changes.

nicholasloveday replied:
Yes, at the time you reviewed this, I was mid-way through the refactor and hadn't touched the notebook. I've rewritten the tutorial now since I removed `threshold_weighted_score` from this pull request.

tennlee reviewed (changes requested):
Looks very much in the right direction, there are just a few things I've noted in my review. Also, I haven't reviewed the tests as yet. Thanks very much for this work.

nicholasloveday commented:
Thanks for that feedback @tennlee - I'm still in the middle of the refactor, so please hold off on anymore feedback until this is ready for review again.

Steph-Chong commented on docs/included.md:
```diff
@@ -117,6 +117,26 @@
+  - [API](api.md#scores.continuous.tw_absolute_error)
+  - [Tutorial](project:./tutorials/Threshold_Weighted_Scores.md)
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+  * - Threshold Weighted Expectile Score
+  - [API](api.md#scores.continuous.tw_expectile_score)
```
```suggestion
* - Threshold Weighted Expectile Score
  - [API](api.md#scores.continuous.tw_expectile_score)
```

Steph-Chong commented on docs/included.md:
```diff
@@ -117,6 +117,26 @@
+  - [API](api.md#scores.continuous.tw_expectile_score)
+  - [Tutorial](project:./tutorials/Threshold_Weighted_Scores.md)
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+  * - Threshold Weighted Huber Loss
+  - [API](api.md#scores.continuous.tw_huber_loss)
```
```suggestion
* - Threshold Weighted Huber Loss
  - [API](api.md#scores.continuous.tw_huber_loss)
```

Steph-Chong commented on docs/included.md:
```diff
@@ -117,6 +117,26 @@
+  - [API](api.md#scores.continuous.tw_huber_loss)
+  - [Tutorial](project:./tutorials/Threshold_Weighted_Scores.md)
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+  * - Threshold Weighted Squared Error
+  - [API](api.md#scores.continuous.tw_squared_error)
```
```suggestion
* - Threshold Weighted Squared Error
  - [API](api.md#scores.continuous.tw_squared_error)
```

Steph-Chong commented on docs/included.md:
```diff
@@ -117,6 +117,26 @@
+  - [API](api.md#scores.continuous.tw_squared_error)
+  - [Tutorial](project:./tutorials/Threshold_Weighted_Scores.md)
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+  * - Threshold Weighted Quantile Score
+  - [API](api.md#scores.continuous.tw_quantile_score)
```
```suggestion
* - Threshold Weighted Quantile Score
  - [API](api.md#scores.continuous.tw_quantile_score)
```

Steph-Chong commented:
> Thanks for that feedback @tennlee - I'm still in the middle of the refactor, so please hold off on anymore feedback until this is ready for review again.

Sorry @nicholasloveday - I didn't see this comment before I made my comments. If I had seen it first, I would not have commented. My apologies.

nicholasloveday commented:
> > Thanks for that feedback @tennlee - I'm still in the middle of the refactor, so please hold off on anymore feedback until this is ready for review again.
> 
> Sorry @nicholasloveday - I didn't see this comment before I made my comments. If I had seen it first, I would not have commented. My apologies.

No problem! I just didn't want people using up their time reviewing something that is only partly done :)

nicholasloveday commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,687 @@
+) -> Tuple[AuxFuncType, AuxFuncType, AuxFuncType]:
+    """
+    Returns the three auxiliary functions g, phi and phi_prime
+    which are used to construct quantile, expectile or Huber mean scoring functions.
```
Added args

nicholasloveday commented:
Okay, I've finished the refactor and responded to that initial feedback. It's now ready for review again thanks @tennlee and @Steph-Chong

nicholasloveday commented:
I should add that you may need to run the tutorials yourself rather than view them on GitHub so that the Plotly figures display

tennlee commented on docs/included.md:
```diff
@@ -117,6 +117,26 @@
+  - [API](api.md#scores.continuous.tw_absolute_error)
+  - [Tutorial](project:./tutorials/Threshold_Weighted_Scores.md)
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+* - Threshold Weighted Expectile Score
```
It seems like the "*" has a space before it that shouldn't be there. I brought it in line with the others in my checkout and that enabled the docs to build.

tennlee commented on tutorials/Tutorial_Gallery.ipynb:
```diff
@@ -57,7 +57,8 @@
     "- [Flip-Flop Index](./Flip_Flop_Index.ipynb)\n",
-    "- [Consistent Scores](./Consistent_Scores.ipynb)"
+    "- [Consistent Scores](./Consistent_Scores.ipynb)\n",
+    "- [Threshold Weighted Scores](./Threshold_Weights_scores.ipynb)"
```
I suggest renaming the notebook file to "Threshold_Weighted_Scores.ipynb" for consistency. There are some links which point to that (e.g. the consistency notebook and included.md) which don't resolve as a result.

tennlee commented:
To get the nice plots in the notebooks to work, I had to install the "ipywidgets" package. I suggest we add it to the "maintainer" dependencies". I also had to instead of using "fig.show()" just end each cell with "fig". This triggers plotly to dynamically detect the right rendering backend for sphinx.

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,739 @@
+    see :py:func:`scores.continuous.consistent_expectile_score`.
+
+    Two types of threshold weighting are supported: rectangular and trapezoidal.
+
```
Please remove the blank line on line 327. This helps with the sphinx rendering of the API docs in this situation (the whens and whys of blank lined, indentation and bullet lists are very nuanced)

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,739 @@
+    see :py:func:`scores.continuous.consistent_quantile_score`.
+
+    Two types of threshold weighting are supported: rectangular and trapezoidal.
+
```
Please remove the blank line on line 408. This helps with the sphinx rendering of the API docs in this situation (the whens and whys of blank lined, indentation and bullet lists are very nuanced)

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,739 @@
+    see :py:func:`scores.continuous.consistent_quantile_score`.
+
+    Two types of threshold weighting are supported: rectangular and trapezoidal.
+
```
Please remove the blank line on line 492. This helps with the sphinx rendering of the API docs in this situation (the whens and whys of blank lined, indentation and bullet lists are very nuanced)

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,739 @@
+    see :py:func:`scores.continuous.consistent_expectile_score`.
+
+    Two types of threshold weighting are supported: rectangular and trapezoidal.
+
```
Please remove the blank line on line 478. This helps with the sphinx rendering of the API docs in this situation (the whens and whys of blank lined, indentation and bullet lists are very nuanced)

tennlee commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,739 @@
+    see :py:func:`scores.continuous.consistent_huber_score`.
+
+    Two types of threshold weighting are supported: rectangular and trapezoidal.
+
```
Please remove the blank line on line 664. This helps with the sphinx rendering of the API docs in this situation (the whens and whys of blank lined, indentation and bullet lists are very nuanced)

nicholasloveday commented:
Thanks for that useful feedback @tennlee - I have pushed up the changes

tennlee commented on tutorials/Tutorial_Gallery.ipynb:
```diff
@@ -57,7 +57,8 @@
     "- [Flip-Flop Index](./Flip_Flop_Index.ipynb)\n",
-    "- [Consistent Scores](./Consistent_Scores.ipynb)"
+    "- [Consistent Scores](./Consistent_Scores.ipynb)\n",
+    "- [Threshold Weighted Scores](./Threshold_Weights_Scores.ipynb)"
```
I think this should link to Threshold_Weighted_Scores.ipynb

nicholasloveday replied:
Yes you're right. I'll update it

Steph-Chong commented on src/scores/continuous/threshold_weighted_impl.py:
```diff
@@ -0,0 +1,734 @@
+    see :py:func:`scores.continuous.consistent_quantile_score`.
+
+    Two types of threshold weighting are supported: rectangular and trapezoidal.
+        - To specify a rectangular weight, set `interval_where_positive=None` and set
```
In readthedocs, back ticks in API docstrings render differently compared to how back ticks in .md files render in readthdocs.

The following comments specifically relate only to API docstrings rendering in readthedocs:
- When you use one back tick on either side of text in API docstrings, readthedocs renders the text as italics, and does not render it as code (in a box, so to speak). For instance,  \`interval_where_positive=None\` is currently rendering as italic text (and not as code, in a box) in readthedocs.
- In order for readthedocs to render text as code (in a box), you need two back ticks on either side of the text. For instance, \``interval_where_positive=None`` should render as code (in a box) in readthedocs
- An exception to this, is when the back ticks are proceeded by :py:func: - in these intances, you only need one back tick, then they render as code (in a box) with a hyperlink. For instance, :py:func:\`scores.continuous.consistent_quantile_score\` currently renders in readthedocs as code (in a box) with a hyperlink.

tennlee replied:
<img width="648" alt="image" src="https://github.com/user-attachments/assets/5dd3c77e-4b87-40c7-80d1-31e6f5c42aaf">

tennlee replied:
The "flexible weighting schemes" link has a py:func
The "set_interval_where_positive" bit has a single backtick and need a double-backtick
This magic only applies to docstrings, other magic applies elsewhere

nicholasloveday replied:
Hopefully I fixed this (I may have missed some)

nicholasloveday commented on docs/included.md:
```diff
@@ -117,6 +117,26 @@
+  - [API](api.md#scores.continuous.tw_squared_error)
+  - [Tutorial](project:./tutorials/Threshold_Weighted_Scores.md)
+  - [Taggart (2022)](https://doi.org/10.1002/qj.4206)
+* - Threshold Weighted Quantile Score
```
Thanks -reordered

tennlee commented:
Thanks for this. I have made a few fixes into the fork and pushed them up. The docs are now building locally so I have pushed the code up to https://scores.readthedocs.io/en/228-documentation-testing-branch/ so we can test it there. 

A question - does it make sense to add mathjax for the new methods, or are they too complex to represent as a formula?

tennlee commented:
"228" on readthedocs looks good to me. Can you check it also and make sure I haven't missed anything and drop a comment here. Also let me know about the mathjax. I think those are the two remaining questions before this can be merged.

nicholasloveday commented:
> Thanks for this. I have made a few fixes into the fork and pushed them up. The docs are now building locally so I have pushed the code up to https://scores.readthedocs.io/en/228-documentation-testing-branch/ so we can test it there.
> 
> A question - does it make sense to add mathjax for the new methods, or are they too complex to represent as a formula?

I did think about it, but decided to leave them out. I think that the equations will be so long that anyone interested in them is better off going to the paper.

tennlee commented:
Here is the link to the list of things to check for on readthedocs: https://scores.readthedocs.io/en/stable/maintainer.html#this-section-covers-checking-the-documentation-renders-properly-in-readthedocs

nicholasloveday commented:
I've gone through the readthedocs and have made several changes. I think that they addressed the issues that you raised. I've also made other changes (e.g., cut down whitespace on the figs in the tutorial)

tennlee commented:
> I've gone through the readthedocs and have made several changes. I think that they addressed the issues that you raised. I've also made other changes (e.g., cut down whitespace on the figs in the tutorial)

Thanks. I haven't reviewed the changes yet, but I have pushed your latest work up to 228. It will take a couple of minutes to build and then you can take a look at how the changes have come through.

Steph-Chong commented:
In the tutorial, it looks like the plots aren't rendering in nbviewer, as viewed here: https://nbviewer.org/urls/scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Threshold_Weighted_Scores.ipynb

Note: sometimes nbviewer when viewed from the link at the top of the readthedocs version (as linked above) of the tutorials renders differently compared to how nbviewer renders on the preview screen while binder is launching. So it is possible, the nbviewer preview on the binder launch screen will render differently. (I haven't tested nbviewr in the binder launch screen, if anyone wants to test it, I think you may need to change the URL to test it from 228 branch - off the top of my head the default binder link is linked to main).

tennlee commented:
> There is maybe an issue with the plots again.
> 
> The plots don't appear to be rendering in:
> 
> 1. readthedocs: https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Threshold_Weighted_Scores.html
> 2. nbviewer - from the link at the top of the readthedocs tutorial page): https://nbviewer.org/urls/scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Threshold_Weighted_Scores.ipynb
> 3. nbviewer - from the preview screen at the bottom of the binder launch page: https://mybinder.org/v2/gh/nci/scores/228-documentation-testing-branch?labpath=tutorials/Threshold_Weighted_Scores.ipynb
> 
> The plots do seem to be rendering in binder itself (once launched).
> 
> Note re. nbviewer: nbviewer can render differently in the two locations listed above.

Looks like all of this is linked to "whatever makes the notebooks work when they are built on my machine". I re-ran the notebook in my environment and pushed the change up to 228. I'm sure we'll get to the root cause eventually, but for now I will just have to re-run the notebooks each time I push to 228.

nicholasloveday commented:
@tennlee I have made minor updates to the tutorial based on @rob-taggart feedback that he provided offline. I didn't regenerate the plots, so they should still display fine in the docs (they display for me when I generate the docs).

Steph-Chong commented on tutorials/Threshold_Weighted_Scores.ipynb:
```diff
@@ -8,7 +8,7 @@
     "For a long time, people have wanted to be able to evaluate the performance of forecasts of extreme events using verification methods that can't be hedged. The threshold weighted scores in `scores` allow us to do this using consistent scoring functions.\n",
     "\n",
-    "Threshold weighted scores allow us to measure the performance of forecasts with a focus on a range of decision thresholds (e.g., performance of temperature forecasts above 40°C) using [consistent scoring functions](./Consistent_Scores.ipynb) and they do not allow hedging. They allow us to measure the performance of predicting extremes or some other range of decision thresholds of interest. T ...
+    "Threshold weighted scores allow us to measure the performance of forecasts with a focus on a range of decision thresholds (e.g., users who key decision thresholds are above 40°C) using [consistent scoring functions](./Consistent_Scores.ipynb) and they do not allow hedging. They allow us to measure the performance of predicting extremes or some other range of decision thresholds of interest.  ...
```
I am a bit confused by the phrase "(e.g., users who key decision thresholds are above 40°C)". Do you mean something along the line of: "users whose key decision thresholds are temperatures above 40°C" or "users with a key decision threshold of temperatures above 40°C"?

nicholasloveday replied:
Thanks, I've added the word "temperature" for clarity

Steph-Chong replied:
This may be me misunderstanding, but is the "who" meant to be a "whose"? 

E.g. "users whose key decision thresholds are temperatures above 40°C"

nicholasloveday replied:
Yes, that's exactly what I meant to write. I'll update it now.

Steph-Chong commented:
I couldn't work out how to leave these as code comments. 

Some very minor feedback about the tutorial header levels (as they currently appear in branch 228 - I didn't rebuild the docs, so apologies if these have already been addressed):

1.  Header "Other Threshold Weighted Scores." has a full stop at the end. I am guessing the full stop isn't meant to be there.
2. Some of the capitalisation of the headers is inconsistent. In particular,
2a. "Rectangular weighting", could consider capitalising "w" in "weighting"
2b. "Trapezoidal weighting", could consider capitalising "w" in "weighting"
2c. "Things to try next", could consider capitalising "To Try Next"

Steph-Chong commented:
In the "Other Threshold Weighted Scores" section, the first sentence says: "In this tutorial, we demonstrated the threshold weighted square error (or twMSE) function."

I think there may be a missing "mean" - as in "threshold weighted mean square error (or twMSE) function."

nicholasloveday commented:
Thanks @Steph-Chong for those suggestions. I've made some updates that address your feedback.

nicholasloveday commented:
> > There is maybe an issue with the plots again.
> > The plots don't appear to be rendering in:
> > 
> > 1. readthedocs: https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Threshold_Weighted_Scores.html
> > 2. nbviewer - from the link at the top of the readthedocs tutorial page): https://nbviewer.org/urls/scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Threshold_Weighted_Scores.ipynb
> > 3. nbviewer - from the preview screen at the bottom of the binder launch page: https://mybinder.org/v2/gh/nci/scores/228-documentation-testing-branch?labpath=tutorials/Threshold_Weighted_Scores.ipynb
> > 
> > The plots do seem to be rendering in binder itself (once launched).
> > Note re. nbviewer: nbviewer can render differently in the two locations listed above.
> 
> Looks like all of this is linked to "whatever makes the notebooks work when they are built on my machine". I re-ran the notebook in my environment and pushed the change up to 228. I'm sure we'll get to the root cause eventually, but for now I will just have to re-run the notebooks each time I push to 228.

I think that the issue isn't the environment. When I run the notebooks in VS code they don't render when I build the docs. When I run the notebooks in a a Jupyter notebook in a web browser, then they do render when I build the docs!

tennlee commented:
I am making comments about the tutorial here, since the file is too large for inline comments through github code review.

Regarding the tutorial, it notes that MSE can be interpreted as putting equal weight on all thresholds. Since it squares the error, it puts exponentially more weight on higher thresholds, so I think this isn't quite right to say. Also, the previous version noted the issue with MSE potentially obscuring certain things.

tennlee commented:
With rectangular weighting, it's unclear if multiple segments/intervals could be supplied, e.g. the user might want to weight anything <= 3 and also anything >= 3 as "1", and "0" between -3 and positive 3. I'm assuming this is currently not possible, but the tutorial could do a better job of explaining what an "interval" is here.

tennlee commented:
The trapezoidal weighting example given in the paragraph "Trapezoidal Weighting" isn't the same as the one presented in code in the example. Consider adding a plot to the "Trapezoidal Weighting" section to show what is described in the text, or consider adjusting the text description to match the later code and say "see below for an example"

tennlee commented:
> With rectangular weighting, it's unclear if multiple segments/intervals could be supplied, e.g. the user might want to weight anything <= 3 and also anything >= 3 as "1", and "0" between -3 and positive 3. I'm assuming this is currently not possible, but the tutorial could do a better job of explaining what an "interval" is here.

Also, how does the user specify greater-than-or-equal to some threshold? i.e. from 3 to infinity.

One option here would be to replace the threshold_where_one with a callable which returned '1' when 1 and 0 when 0. 

The nice thing about the current approach is that it's simpler for non-coders who might be less comfortable making up functions on the fly, and also to provide some guidance on the most common things to do. One way to support both approaches would be to be support some optional argument which would allow a function to be passed instead.

I also think we might need to explain why these weighting parameters are being passed in rather than using the already suitable "weights" argument to specify the weights. For example, why are we not using functions like weights_where_one or trap_weights to calculate a suitable weights array instead? Part of the answer is convenience and guidance for users, but I think it could be confusing to people who are thinking about the weights. Do the functions treat the weights in `weights` differently from the weightings which result frorm the various weighting parameters that are passed in?

tennlee commented:
It might be interesting to add an example for a variable which has both positive and negative values, either where the interval of interest is around the mean, or where it's on the extremes. It's not clear to me how this can be used on something like temperature, where both very hot and very cold extremes are both of similar (although potentially non-equal) interest. I really like the examples that are currently included, and I'm a bit worried about overcomplicating the tutorial, but I want to make sure this approach is suitable for that situation and an example would both put the implementation to the test and demonstrate appropriate practise. 

Examples could include for example human comfort ranges for temperature or humidity, agriculture (where both heat and frost/ice could be problems), or operating equipment like trains where again there would be a temperature range.

rob-taggart commented:
> I am making comments about the tutorial here, since the file is too large for inline comments through github code review.
> 
> Regarding the tutorial, it notes that MSE can be interpreted as putting equal weight on all thresholds. Since it squares the error, it puts exponentially more weight on higher thresholds (since multiplicative biases at higher thresholds will consequentially have higher absolute errors which are then squared), so I think this isn't quite right to say. Also, the previous version noted the issue with MSE potentially obscuring certain things.

@tennlee , they way you have put it is not quite correct.

Take the following example. My decision threshold is 40C. If tomorrow's forecast exceeds 40C, I will tell the workers on my construction site not to come into work tomorrow. Now consider the case where the forecast for tomorrow was 20C and the observations was 10C. That is a large error, and the squared error is even larger (100C^2). However, I still made the correct decision on the basis of the forecast to not close the construction site. So the size of the penalty can be independent of whether the forecast was good enough for people whose decision thresholds are 40C or higher. In this case, threshold-weighted squared error actually give a penalty of 0.

What Nick says is correct. MSE can be interpreted as putting equal weight on all thresholds. The Ehm 2016 paper explains why. Happy to talk about this in another forum rather than a gitlab comment

rob-taggart commented:
@nicholasloveday , one of @tennlee comments about weights makes me wonder whether:

1. there is a weight argument in the threshold weighted scores, which applies specified weights to each of the individual score prior to calculating the mean across some dimensions; and
2. how this kind of weight differs from threshold-weighting, and whether that is clear in the documentation.

I noticed in your docstrings that you talk about weights where one, etc, which is the second type of weight listed above, not the first.

To avoid confusion, you may need to talk consistently about threshold-weights (and explain that this is related to decision thresholds) to distinguish this from the other type of weighting. Both in the tutorial and in the docstrings for your functions

tennlee commented:
> > I am making comments about the tutorial here, since the file is too large for inline comments through github code review.
> > Regarding the tutorial, it notes that MSE can be interpreted as putting equal weight on all thresholds. Since it squares the error, it puts exponentially more weight on higher thresholds (since multiplicative biases at higher thresholds will consequentially have higher absolute errors which are then squared), so I think this isn't quite right to say. Also, the previous version noted the issue with MSE potentially obscuring certain things.
> 
> @tennlee , they way you have put it is not quite correct.
> 
> Take the following example. My decision threshold is 40C. If tomorrow's forecast exceeds 40C, I will tell the workers on my construction site not to come into work tomorrow. Now consider the case where the forecast for tomorrow was 20C and the observations was 10C. That is a large error, and the squared error is even larger (100C^2). However, I still made the correct decision on the basis of the forecast to not close the construction site. So the size of the penalty can be independent of whether the forecast was good enough for people whose decision thresholds are 40C or higher. In this case, threshold-weighted squared error actually give a penalty of 0.
> 
> What Nick says is correct. MSE can be interpreted as putting equal weight on all thresholds. The Ehm 2016 paper explains why. Happy to talk about this in another forum rather than a gitlab comment

Thanks, I've followed up offline.

My comments aren't about what's right and wrong, but about whether things are clear when introduced to a new reader with a general mathematical understanding of errors but without prior knowledge of decision thresholds. i.e. to help people avoid confusion about the exact disagreement being expressed in the comment.

tennlee commented:
> @nicholasloveday , one of @tennlee comments about weights makes me wonder whether:
> 
>     1. there is a weight argument in the threshold weighted scores, which applies specified weights to each of the individual score prior to calculating the mean across some dimensions; and
> 
>     2. how this kind of weight differs from threshold-weighting, and whether that is clear in the documentation.
> 
> 
> I noticed in your docstrings that you talk about weights where one, etc, which is the second type of weight listed above, not the first.
> 
> To avoid confusion, you may need to talk consistently about threshold-weights (and explain that this is related to decision thresholds) to distinguish this from the other type of weighting. Both in the tutorial and in the docstrings for your functions

Yes, spot on.

tennlee commented:
> Take the following example. My decision threshold is 40C. If tomorrow's forecast exceeds 40C, I will tell the workers on my construction site not to come into work tomorrow. Now consider the case where the forecast for tomorrow was 20C and the observations was 10C. That is a large error, and the squared error is even larger (100C^2). However, I still made the correct decision on the basis of the forecast to not close the construction site. So the size of the penalty can be independent of whether the forecast was good enough for people whose decision thresholds are 40C or higher. In this case, threshold-weighted squared error actually give a penalty of 0

This would be a nice example to include.

nicholasloveday commented:
> The trapezoidal weighting example given in the paragraph "Trapezoidal Weighting" isn't the same as the one presented in code in the example. Consider adding a plot to the "Trapezoidal Weighting" section to show what is described in the text, or consider adjusting the text description to match the later code and say "see below for an example"

Added a text description in the trapezoidal weighting paragraph. I added rather than replaced as it is important to show how trapezoidal weights can be on either one side or both sides of the rectangular weights.



> Take the following example. My decision threshold is 40C. If tomorrow's forecast exceeds 40C, I will tell the workers on my construction site not to come into work tomorrow. Now consider the case where the forecast for tomorrow was 20C and the observations was 10C. That is a large error, and the squared error is even larger (100C^2). However, I still made the correct decision on the basis of the forecast to not close the construction site. So the size of the penalty can be independent of whether the forecast was good enough for people whose decision thresholds are 40C or higher. In this case, threshold-weighted squared error actually give a penalty of 0
> This would be a nice example to include.

I've added a modified example that lends itself more to a range of decision thresholds rather than a single decision threshold


> I noticed in your docstrings that you talk about weights where one, etc, which is the second type of weight listed above, not the first.
> 
> To avoid confusion, you may need to talk consistently about threshold-weights (and explain that this is related to decision thresholds) to distinguish this from the other type of weighting. Both in the tutorial and in the docstrings for your functions

I have updated the tutorial and docstrings to clarify this.

> With rectangular weighting, it's unclear if multiple segments/intervals could be supplied, e.g. the user might want to weight anything <= 3 and also anything >= 3 as "1", and "0" between -3 and positive 3. I'm assuming this is currently not possible, but the tutorial could do a better job of explaining what an "interval" is here.
>  It might be interesting to add an example for a variable which has both positive and negative values, either where the interval of interest is around the mean, or where it's on the extremes. It's not clear to me how this can be used on something like temperature, where both very hot and very cold extremes are both of similar (although potentially non-equal) interest. I really like the examples that are currently included, and I'm a bit worried about overcomplicating the tutorial, but I want to make sure this approach is suitable for that situation and an example would both put the implementation to the test and demonstrate appropriate practise.
> 
> Examples could include for example human comfort ranges for temperature or humidity, agriculture (where both heat and frost/ice could be problems), or operating equipment like trains where again there would be a temperature range.

Added an example in the tutorial. Also added an additional thing to try in the "things to try next" section.
